File: src/main/java/Leetcode/Linked/Train.java
package Leetcode.Linked;

import java.util.TreeMap;

public class Train {
    public static void main(String[] args) {

    }

    /**
     * 203. 移除链表元素(循环)
     * @param head
     * @param val
     * @return
     */
    public Leetcode.Linked.ListNode removeElements(Leetcode.Linked.ListNode head, int val) {
        Leetcode.Linked.ListNode dummyHead = new Leetcode.Linked.ListNode(-1);
        dummyHead.next = head;
        Leetcode.Linked.ListNode prev = dummyHead;
        while (prev.next != null) {
            if (prev.next.val == val) {
                prev.next = prev.next.next;
            }else{
                prev = prev.next;
            }
        }
        return dummyHead.next;
    }

    /**
     * 203. 移除链表元素(递归)
     * @param head
     * @param val
     * @return
     */
    public Leetcode.Linked.ListNode removeElements2(Leetcode.Linked.ListNode head, int val) {
        if (head == null) {return null;}
        head.next = removeElements2(head.next,val);
        return head.val == val ? head.next : head;
    }

    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    /**
     * 21. 合并两个有序链表
     * @param l1
     * @param l2
     * @return
     */
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode cur = dummyHead;

        while(l1 != null && l2 != null){
            if(l1.val<= l2.val){
                cur.next = l1;
                cur = cur.next;
                l1 = l1.next;
            }else{
                cur.next = l2;
                cur = cur.next;
                l2 = l2.next;
            }
        }
        if(l1 == null){
            cur.next = l2;
        }else{
            cur.next = l1;
        }
        return dummyHead.next;
    }

    /**
     * 83. 删除排序链表中的重复元素
     * @param head
     * @return
     */
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null || head.next == null){return null;}
        head.next = deleteDuplicates(head.next);
        if(head.val == head.next.val){head = head.next;}
        return head;
    }

    /**
     * 876. 链表的中间结点
     * @param head
     * @return
     */
    public ListNode middleNode(ListNode head) {
        ListNode dummyHead = head;
        while(dummyHead != null && dummyHead.next != null){
            dummyHead = dummyHead.next.next;
            head = head.next;
        }
        return head;
    }

    /**
     * 237. 删除链表中的节点
     * @param node
     */
    public void deleteNode(ListNode node) {
        node.val = node.next.val;
        node.next = node.next.next;
    }

    /**
     * 206. 反转链表
     * @param head
     * @return
     */
    public ListNode reverseList(ListNode head) {
        ListNode newHead = null;
        while (head != null){
            ListNode temp = head.next;
            head.next = newHead;
            newHead = head;
            head = temp;
        }
        return newHead;
    }


}

File: src/main/scala/com/Leetcode/Linked/Train.scala
package com.Leetcode.Linked




object Train {

  def main(args: Array[String]): Unit = {
    var root = new ListNode(1)
    var h1 = new ListNode(2)
    root.next = h1
    var h2 = new ListNode(6)
    h1.next = h2
    var h3 = new ListNode(3)
    h2.next = h3
    var h4 = new ListNode(4)
    h3.next = h4
    var h5 = new ListNode(5)
    h4.next = h5
    var h6 = new ListNode(6)
    h5.next = h6
    removeElements2(root,6)
  }

  class ListNode(var _x: Int = 0) {
    var next: ListNode = null
    var x: Int = _x
  }

  /**
    * 203. 移除链表元素（循环）
    * @param head
    * @param `val`
    * @return
    */
  def removeElements(head: ListNode, `val`: Int): ListNode = {
    var dummyHead:ListNode  = new ListNode(-1)
    dummyHead.next = head

    var prev = dummyHead
    while (prev.next != null) {
      if (prev.next.x == `val`) {
        prev.next = prev.next.next
      }else{
        prev = prev.next
      }
    }
    dummyHead.next
  }

  /**
    * 203. 移除链表元素(递归)
    * @param head
    * @param `val`
    * @return
    */
  def removeElements2(head: ListNode, `val`: Int): ListNode = {
    if (head == null ) {return null}
    var res = removeElements2(head.next,`val`)
    if (head.x == `val`) {res}
    else{
      head.next = res
      head
    }
  }

  class ListNode2(var _x: Int = 0) {
    var next: ListNode2 = null
    var x: Int = _x
  }

  /**
    * 21. 合并两个有序链表
    * @param l1
    * @param l2
    * @return
    */
  def mergeTwoLists(l1: ListNode2, l2: ListNode2): ListNode2 = {
    val dummyHead = new ListNode2()
    var p1 = l1
    var p2 = l2
    var cur = dummyHead
    while(p1 != null && p2 != null){
      if (p1.x <= p2.x) {
        cur.next = p1
        cur = cur.next
        p1 = p1.next
      }else{
        cur.next = p2
        cur = cur.next
        p2 = p2.next
      }
    }
    if (p1 == null) {cur.next = p2}
    else{cur.next = p1}
    dummyHead.next
  }

  /**
    * 83. 删除排序链表中的重复元素
    * @param head
    * @return
    */
  def deleteDuplicates(head: ListNode): ListNode = {
    if(head == null || head.next == null){
      return head
    }
    head.next = deleteDuplicates(head.next)
    if(head.x == head.next.x){head.next}
    else {head}
  }

  /**
    * 876. 链表的中间结点
    * @param head
    * @return
    */
  def middleNode(head: ListNode): ListNode = {
    var dummyHead = head
    var cur = head
    while(dummyHead != null && dummyHead.next != null){
      dummyHead = dummyHead.next.next
      cur = cur.next
    }
    cur
  }

//  /**
//    * 206. 反转链表
//    *
//    * @param head
//    * @return
//    */
//  def reverseList(head: ListNode): ListNode = {
//
//  }

}
